src/constraint_functions.jl

```diff
@@ -93,4 +91,400 @@ function EMB.constraints_level_aux(m, n::HydroStorage, 𝒯, 𝒫, modeltype::En
     )
 end
 
-#! format: on
+"""
+    build_hydro_reservoir_vol_constraints(m, n::HydroReservoir, c::Constraint, 𝒯)
+
+Create minimum/maximum/schedule volume constraints for a [`HydroReservoir`](@ref) node. The
+`Constraint{T}` can have types `T <: AbstractConstraintType` that defines the direction of
+the constraint.
+
+Penalty variables are included unless the  penalty value is not set or `Inf`.
+"""
+function build_hydro_reservoir_vol_constraints(m, n::HydroReservoir, c::Constraint{MinConstraintType}, 𝒯)
+    p = storage_resource(n)
+    @constraint(m, [t ∈ 𝒯; is_active(c, t) & has_penalty(c, t)],
+        m[:stor_level][n, t] + m[:rsv_penalty_up][n, t, p] ≥ EMB.capacity(EMB.level(n), t) * value(c, t))
+    @constraint(m, [t ∈ 𝒯; is_active(c, t) & !has_penalty(c, t)],
+        m[:stor_level][n, t] ≥ EMB.capacity(EMB.level(n), t) * value(c, t))
+end
+function build_hydro_reservoir_vol_constraints(m, n::HydroReservoir, c::Constraint{MaxConstraintType}, 𝒯)
+    p = storage_resource(n)
+    @constraint(m, [t ∈ 𝒯; is_active(c, t) & has_penalty(c, t)],
+        m[:stor_level][n, t] - m[:rsv_penalty_down][n, t, p] ≤ EMB.capacity(EMB.level(n), t) * value(c, t))
+    @constraint(m, [t ∈ 𝒯; is_active(c, t) & !has_penalty(c, t)],
+        m[:stor_level][n, t] ≤ EMB.capacity(EMB.level(n), t) * value(c, t))
+end
+function build_hydro_reservoir_vol_constraints(m, n::HydroReservoir, c::Constraint{ScheduleConstraintType}, 𝒯)
+    p = storage_resource(n)
+    @constraint(m, [t ∈ 𝒯; is_active(c, t) & has_penalty(c, t)],
+        m[:stor_level][n, t] + m[:rsv_penalty_up][n, t, p] - m[:rsv_penalty_down][n, t, p] ==
+        EMB.capacity(EMB.level(n), t) * value(c, t))
+    for t ∈ 𝒯
+        if is_active(c, t) & !has_penalty(c, t)
+            JuMP.fix(m[:stor_level][n, t], EMB.capacity(EMB.level(n), t) * value(c, t))
```

Comment:
The direct utilization of `EMB.capacity` here implies that we cannot use investment models for the individual nodes. This is not a major issue, but must be highlighted both in the documentation and checked in the checks through a function like

```julia
function EMB.check_node_data(
    n::HydroUnit,
    data::InvestmentData,
    𝒯,
    modeltype::AbstractInvestmentModel,
    check_timeprofiles::Bool,
)

    @assert_or_log(
        !has_investment(n),
        "`InvestmentData` is not allowed for `HydrogenStorage` nodes."
    )
end
```

This function would be located in an extension.

Reply:
I can do it.

Reply:
Solved in https://github.com/EnergyModelsX/EnergyModelsRenewableProducers.jl/pull/35/commits/1b69494b33ed53db959eabe0997357e5034ae5c3